## Theory Questions

### 1. What is the difference between a function and a method in Python?

A function is defined using the def keyword and is not associated with any object.It can exist independently and be called using its name. A method is a function that is associated with an object. It is called using the dot (.) operator on an object. The first parameter is usually self ,referring to the object itself.

### 2. Explain the concept of function arguments and parameters in Python.

Parameters are variables listed in a function definition. They act as placeholders for values that will be passed to the function. Arguments are the actual values passed to the function when it is called.

### 3. What are the different ways to define and call a function in Python?

In Python, functions can be defined and called in several ways depending on the use case. 

1. Standard function

In [7]:
def greet(name):
    return f"Hello, {name}!"

In [8]:
func = greet
print(func("Eve")) 

Hello, Eve!


2. Anonymous Function

In [9]:
greet = lambda name: f"Hello, {name}!"

In [10]:
message = greet("Alice")
print(message)

Hello, Alice!


3. Function with Default Arguments

In [11]:
def greet(name="Aswini"):
    return f"Hello, {name}!"

In [12]:
greet()

'Hello, Aswini!'

4. Function with Variable-Length Arguments

In [13]:
def add(*args):
    return sum(args)

In [14]:
print(add(1, 2, 3))

6


In [15]:
def show_info(**kwargs):
    for key, value in kwargs.items():
        print(key, ":", value)

In [16]:
show_info(name="Alice", age=25)

name : Alice
age : 25


### 4. What is the purpose of the `return` statement in a Python function?

The return statement in Python is used to Exit a function immediately, send a result back to the caller of the function.

In [17]:
def add(a, b):
    return a + b

result = add(3, 5)
print(result)  # Output: 8


8


### 5. What are iterators in Python and how do they differ from iterables?

An iterator is an object in Python that allows traversing through a sequence one element at a time using the built-in next() function. To qualify as an iterator, an object must implement two methods:__iter__() returns the iterator object itself, __next__() returns the next item or raises StopIteration when done. An iterable is any object capable of returning its members one at a time using an iterator. It must implement the __iter__() method. Common examples of iterables are Lists, Tuples, Strings, Sets, Ditionary.

### 6. Explain the concept of generators in Python and how they are defined.

Generators are a special type of iterable in Python that yield values one at a time, on demand, using the yield keyword instead of return. They are used to produce sequences efficiently, especially when dealing with large datasets or streams of data. 

Generator functions are defined like a normal function but use yield instead of return. Each time yield is called, the function pauses its state and resumes from the same point the next time it's called.

In [18]:
def count_up_to(n):
    i = 1
    while i <= n:
        yield i
        i += 1

gen = count_up_to(3)
print(next(gen))  

1


Generator Expressions are defined similar to list comprehensions, but use parentheses instead of square brackets.

In [19]:
squares = (x*x for x in range(5))

In [20]:
print(next(squares))

0


### 7. What are the advantages of using generators over regular functions?

Generators are a special type of function that yield values one at a time using the yield keyword, instead of returning all values at once like regular functions. Its memory efficient and useful when dealing with large datasets or infinite sequences.Generators produce values only when needed, not all at once.This improves performance and startup time, especially when not all results are needed.

In [21]:
def gen_nums():
    for i in range(10):
        yield i

def square(nums):
    for n in nums:
        yield n * n

result = square(gen_nums())

### 8. What is a lambda function in Python and when is it typically used?

A lambda function in Python is a small, anonymous function defined using the lambda keyword instead of def. It is generally used for creating throwaway functions. 

In [22]:
add = lambda x, y: x + y
print(add(3, 5))

8


### 9. Explain the purpose and usage of the `map()` function in Python.

The map() function is used to apply a specific function to each item in an iterable (like a list, tuple, etc.) and returns a new iterable (map object) with the results.

In [23]:
numbers = [1, 2, 3, 4]
squares = map(lambda x: x ** 2, numbers)
print(list(squares))

[1, 4, 9, 16]


### 10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

map() applies a function to each element of an iterable.filter() filters elements of an iterable based on a Boolean condition. reduce() applies a function cumulatively to the elements of an iterable, reducing it to a single value. reduce() returns a single value.

### 11. Using pen & Paper write the internal mechanism for sum operation using reduce function on this given list: [47,11,42,13]

In [24]:
from functools import reduce

numbers = [47,11,42,13]
result = reduce(lambda x,y:x+y, numbers)
print(result)

113


First call -> x = 47 ,y = 11 -> x+y = 58 , 
second call -> x = 58, y = 42 -> x+y = 100 , 
Third call -> x=100, y= 13 -> x+y = 113

## Practical Questions:

### 1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

In [25]:
def sum_even_numbers(numbers):
    even_sum = 0
    for num in numbers:
        if num % 2 == 0:
            even_sum += num
    return even_sum


In [26]:
my_list = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(my_list)
print("Sum of even numbers:", result)  # Output: 12


Sum of even numbers: 12


### 2. Create a Python function that accepts a string and returns the reverse of that string.

In [27]:
def reverse_string(text):
    return text[::-1]

In [28]:
input_text = "Hello, World!"
reversed_text = reverse_string(input_text)
print(reversed_text)

!dlroW ,olleH


### 3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [29]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

In [30]:
input_list = [1, 2, 3, 4, 5]
squared_list = square_numbers(input_list)
print(squared_list)

[1, 4, 9, 16, 25]


### 4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [31]:
def is_prime(n):
    if n < 2 or n > 200:
        return False  # Not in the valid range
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

In [32]:
for num in range(1, 201):
    if is_prime(num):
        print(f"{num} is a prime number")

2 is a prime number
3 is a prime number
5 is a prime number
7 is a prime number
11 is a prime number
13 is a prime number
17 is a prime number
19 is a prime number
23 is a prime number
29 is a prime number
31 is a prime number
37 is a prime number
41 is a prime number
43 is a prime number
47 is a prime number
53 is a prime number
59 is a prime number
61 is a prime number
67 is a prime number
71 is a prime number
73 is a prime number
79 is a prime number
83 is a prime number
89 is a prime number
97 is a prime number
101 is a prime number
103 is a prime number
107 is a prime number
109 is a prime number
113 is a prime number
127 is a prime number
131 is a prime number
137 is a prime number
139 is a prime number
149 is a prime number
151 is a prime number
157 is a prime number
163 is a prime number
167 is a prime number
173 is a prime number
179 is a prime number
181 is a prime number
191 is a prime number
193 is a prime number
197 is a prime number
199 is a prime number


### 5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [33]:
class Fibonacci:
    def __init__(self, n):
        self.n = n               
        self.count = 0           
        self.a, self.b = 0, 1    

    def __iter__(self):
        return self              

    def __next__(self):
        if self.count >= self.n:
            raise StopIteration  
        if self.count == 0:
            self.count += 1
            return 0
        elif self.count == 1:
            self.count += 1
            return 1
        else:
            fib = self.a + self.b
            self.a, self.b = self.b, fib
            self.count += 1
            return fib


In [34]:
fib = Fibonacci(10)
for num in fib:
    print(num)

0
1
1
2
3
5
8
13
21
34


### 6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [35]:
def powers_of_two(max_exponent):
    for exp in range(max_exponent + 1):
        yield 2 ** exp

In [36]:
for num in powers_of_two(5):
    print(num)

1
2
4
8
16
32


### 7. Implement a generator function that reads a file line by line and yields each line as a string.

In [37]:
def read_file_lines(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line.rstrip('\n') 

In [38]:
for line in read_file_lines(r"C:\Users\atkco\Documents\PW\sample1.txt.txt"):
    print(line)

Hello, World!
This is a test file.
It contains multiple lines.
Each line will be read and printed.


### 8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [39]:
data = [(1, 4), (3, 1), (5, 2), (2, 3)]
sorted_data = sorted(data, key=lambda x: x[1])

print(sorted_data)

[(3, 1), (5, 2), (2, 3), (1, 4)]


### 9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [40]:
celsius = [0, 20, 37, 100]

fahrenheit = list(map(lambda c: (c * 9/5) + 32, celsius))

print("Celsius:", celsius)
print("Fahrenheit:", fahrenheit)


Celsius: [0, 20, 37, 100]
Fahrenheit: [32.0, 68.0, 98.6, 212.0]


### 10. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [41]:
def remove_vowels(input_str):
    vowels = 'aeiouAEIOU'
    return ''.join(filter(lambda char: char not in vowels, input_str))

text = "Hello, World!"
result = remove_vowels(text)
print("Original:", text)
print("Without vowels:", result)

Original: Hello, World!
Without vowels: Hll, Wrld!


### 11. Imagine an accounting routine used in a book shop. It works on a list with sublists. Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the product of the price per item and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €. Write a Python program using lambda and map.

In [42]:
orders = [
    [34587, 4, 40.95],
    [98762, 1, 100.00],
    [77226, 5, 8.00],
    [88112, 3, 15.00]
]

# Using map and lambda
final_orders = list(map(
    lambda order: (order[0], order[1] * order[2] if order[1] * order[2] >= 100 else order[1] * order[2] + 10),
    orders
))

print(final_orders)


[(34587, 163.8), (98762, 100.0), (77226, 50.0), (88112, 55.0)]
